clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1005,10 +1006,10 @@ class SyclKernelFieldChecker
   // Return true if not copyable, false if copyable.
   bool checkNotCopyableToKernel(const FieldDecl *FD, const QualType &FieldTy) {
     if (FieldTy->isArrayType()) {
-      if (const auto *CAT = dyn_cast<ConstantArrayType>(FieldTy)) {
-        QualType ET = CAT->getElementType();
-        return checkNotCopyableToKernel(FD, ET);
-      }
+      QualType FCT = FieldTy.getUnqualifiedType().getCanonicalType();
+      if (isa<ConstantArrayType>(FCT))
```

Comment:
The above suggestion allows you to do what you were doing above here as well.

Reply:
getAsConstantArrayType is an ASTContext API. To use it, the Visitor functions will need to get the context from the handlers. The context is not currently available outside the handler classes.

Reply:
Unfortunately it is really the only right way to do this... The way above loses context in a few ways that it shouldn't.

I think the visitor functions themselves should probably have a reference to Sema, and the handlers can capture what they need.  

IMO, the visitors likely should all belong to a class.  I'll submit a patch on which you can rebase in order to get this info.